File: secure_examples.py
#!/usr/bin/env python3
# secure_examples.py
# Example of safe parameterized queries in Python (sqlite3). Works in Termux with 'python' package.

import sqlite3

def init_db(path=":memory:"):
    conn = sqlite3.connect(path)
    cur = conn.cursor()
    cur.execute("CREATE TABLE IF NOT EXISTS users(id INTEGER PRIMARY KEY, username TEXT, email TEXT)")
    conn.commit()
    return conn

def insert_user(conn, username, email):
    # Safe: parameterized query (no string concatenation)
    cur = conn.cursor()
    cur.execute("INSERT INTO users (username, email) VALUES (?, ?)", (username, email))
    conn.commit()

def find_user_by_username(conn, username):
    # Safe parameterized SELECT
    cur = conn.cursor()
    cur.execute("SELECT id, username, email FROM users WHERE username = ?", (username,))
    return cur.fetchall()

if __name__ == "__main__":
    c = init_db()
    insert_user(c, "alice", "alice@example.org")
    print(find_user_by_username(c, "alice"))

File: secure_examples.php
<?php
// secure_examples.php
// Example of safe parameterized queries in PHP using PDO.

$dsn = "sqlite::memory:"; // or "mysql:host=127.0.0.1;dbname=test"
try {
    $pdo = new PDO($dsn);
    $pdo->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);
    $pdo->exec("CREATE TABLE IF NOT EXISTS users (id INTEGER PRIMARY KEY, username TEXT, email TEXT)");
} catch (PDOException $e) {
    echo "DB error: " . $e->getMessage();
    exit;
}

function insertUser($pdo, $username, $email) {
    $stmt = $pdo->prepare("INSERT INTO users (username, email) VALUES (:username, :email)");
    $stmt->execute([':username' => $username, ':email' => $email]);
}

function findUserByUsername($pdo, $username) {
    $stmt = $pdo->prepare("SELECT id, username, email FROM users WHERE username = :username");
    $stmt->execute([':username' => $username]);
    return $stmt->fetchAll(PDO::FETCH_ASSOC);
}

insertUser($pdo, 'bob', 'bob@example.org');
print_r(findUserByUsername($pdo, 'bob'));
?>